app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostActivity.java

```diff
@@ -0,0 +1,517 @@
+            break;
+        }
+      }
+    });
+  }
+
+  //Advice user to turn on hotspot manually for API<26
+  void startHotspotDialog() {
+    AlertDialog.Builder builder = new AlertDialog.Builder(this, dialogStyle());
+
+    builder.setPositiveButton(getString(R.string.go_to_settings_label), (dialog, id) -> {
+      setupWifiSettingsIntent();
+    });
+
+    builder.setNeutralButton(getString(R.string.hotspot_dialog_neutral_button), (dialog, id) -> {
+      //TO DO: START SERVER WITHIN THE SERVICE.
+      //Adding a handler because sometimes hotspot can take time to turn on.
+      //TO DO: Add a progress dialog instead of handler
+      ProgressDialog progressDialog =
+          ProgressDialog.show(this, getString(R.string.progress_dialog_starting_server), "",
+              true);
+      progressDialog.show();
+      final Handler handler = new Handler();
+      handler.postDelayed(new Runnable() {
+        @Override
+        public void run() {
+          progressDialog.dismiss();
+          startService(ACTION_START_SERVER);
+        }
+      }, 7000);
+    });
+
+    builder.setTitle(getString(R.string.hotspot_dialog_title));
+    builder.setMessage(
+        getString(R.string.hotspot_dialog_message)
+    );
+    AlertDialog dialog = builder.create();
+    dialog.show();
+  }
+
+  void startService(String ACTION) {
+    serviceIntent.setAction(ACTION);
+    this.startService(serviceIntent);
+  }
+
+  void mobileDataDialog() {
+    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+      AlertDialog.Builder builder = new AlertDialog.Builder(this, dialogStyle());
+
+      builder.setPositiveButton(this.getString(R.string.yes),
+          (dialog, id) -> openDataUsageActivity());
+      builder.setNegativeButton((android.R.string.no), (dialog, id) -> {
+        startHotspotDialog();
+      });
+      builder.setTitle(this.getString(R.string.mobile_data_enabled));
+      builder.setMessage(
+          this.getString(R.string.mobile_data_message) + "\n" + this.getString(
+              R.string.mobile_data_message_confirmation)
+      );
+      builder.setCancelable(false);
+      AlertDialog dialog = builder.create();
+      dialog.show();
+    }
+  }
+
+  @Override public void serverStarted(String ip) {
+    this.ip = ip;
+    serverTextView.setText(getString(R.string.server_started_message) + " " + this.ip);
+    startServerButton.setText(getString(R.string.stop_server_label));
+    startServerButton.setBackgroundColor(getResources().getColor(R.color.stopServer));
+    isServerStarted = true;
+  }
+
+  @Override public void serverStopped() {
+    serverTextView.setText(getString(R.string.server_textview_default_message));
+    startServerButton.setText(getString(R.string.start_server_label));
+    startServerButton.setBackgroundColor(getResources().getColor(R.color.greenTick));
+    isServerStarted = false;
+  }
+
+  @Override public void serverFailed() {
+    String message = String.format(getString(R.string.server_failed_message));
```

Comment:
Quite strange to call String.format and not supply any format args. also Toasts can take resides too

Reply:
It was throwing error otherwise. Since in this string message I've used a `.` in the message which was replaced by some special characters in the background.

Reply:
I would be very surprised if that was the case.
```
Toast.makeText(this, R.string.server_failed_message, Toast.LENGTH_LONG).show();
```
try this and then please report the error here if it presents itself

Reply:
Passing `R.string.server_failed_message` works fine.
But passing `getString(R.string.server_failed_message)` throws this error:

> Format string 'server_failed_message' is not a valid format string so it should not be passed to String.format

Reply:
That is mental, there are `.` in loads of strings in this project. 
https://stackoverflow.com/a/40767291 this seems like a more likely scenario?

Reply:
Fixed in ca12e1791026832b17dfd4bc412faeed84f11a8b